compiler/rustc_const_eval/src/interpret/intern.rs

```diff
@@ -126,6 +131,12 @@ fn intern_shallow<'rt, 'mir, 'tcx, M: CompileTimeMachine<'mir, 'tcx, const_eval:
             // Just making sure we are not "upgrading" an immutable allocation to mutable.
             assert_eq!(alloc.mutability, Mutability::Mut);
         }
+        // It will get interned without an ID as the return value of the `eval_static_initializer` query.
+        if let Some(GlobalAlloc::Static(_)) = tcx.try_get_global_alloc(alloc_id) {
+            // We change the `MemoryKind` to permit validation to read from the allocation.
+            ecx.memory.alloc_map.insert(alloc_id, (MemoryKind::Stack, alloc));
+            return None;
```

Comment:
This feels sketchy and ad-hoc even though it makes perfect sense. I think it shouldn't happen during interning; whoever sets up this allocation should also then, after the interpreter is returned and before interning is called, patch up the allocation kind.

Reply:
It is, this is just the minimal change to get the working and reviewable result. I will add more commits to make it less ad-hoc, but it's gonna be a larger change

Reply:
My comments above still seem to apply, this is doing some wild back-and-forth of allocations; why should the interner be in charge of doing that?

Reply:
I have tried a few things:

* make the interner not remove it
* remove it before the interner in `eval_static_initializer`
* remove it in the interner and return the alloc from the interner

None of them make the code any nicer. The interner is in charge of this, because we're still doing the type based interning. After https://github.com/rust-lang/rust/pull/119044 we won't ever have to look at an allocation twice, so we can just remove it at the start of interning and reinstate it at the end. We still need to remove it so the interner doesn't think it hasn't interned it yet (that's how it decides that after all).